src/plugins/NEMD/MettDeamon.cpp

```diff
@@ -0,0 +1,2048 @@
+
+	double scale_vz = sqrt(numSamples*T/sum_v2zi);
+	for(std::vector<double>::iterator it = vzi.begin(); it != vzi.end(); ++it)
+		(*it) *= scale_vz;
+
+	// update v2 vectors
+	for(uint64_t i=0; i<numSamples; i++)
+	{
+		double vx = vxi.at(i);
+		double vy = vyi.at(i);
+		double vz = vzi.at(i);
+
+		v2xi.at(i) = vx*vx;
+		v2yi.at(i) = vy*vy;
+		v2zi.at(i) = vz*vz;
+	}
+
+	// calc ekin again
+	sum_v2xi = std::accumulate(v2xi.begin(), v2xi.end(), 0.0);
+	sum_v2yi = std::accumulate(v2yi.begin(), v2yi.end(), 0.0);
+	sum_v2zi = std::accumulate(v2zi.begin(), v2zi.end(), 0.0);
+
+	v2x = sum_v2xi * dInvNumSamples;
+	v2y = sum_v2yi * dInvNumSamples;
+	v2z = sum_v2zi * dInvNumSamples;
+
+	cout << "scale x,y,z = " << scale_vx << "," << scale_vy << "," << scale_vz << endl;
+	cout << "v2x,v2y,v2z = " << v2x << "," << v2y << "," << v2z << endl;
+	// <-- EKIN
+
+	// calc drift again
+	sum_vxi = std::accumulate(vxi.begin(), vxi.end(), 0.0);
+	sum_vyi = std::accumulate(vyi.begin(), vyi.end(), 0.0);
+	sum_vzi = std::accumulate(vzi.begin(), vzi.end(), 0.0);
+
+	vxd = sum_vxi * dInvNumSamples;
+	vyd = sum_vyi * dInvNumSamples;
+	vzd = sum_vzi * dInvNumSamples;
+
+	cout << "vxd,vyd,vzd = " << vxd << "," << vyd << "," << vzd << endl;
+}
+
+MettDeamon::MettDeamon() :
+		_reservoir(nullptr),
+		_bIsRestart(false),
+		_bInitFeedrateLog(false),
+		_dAreaXZ(0.),
+		_dInvDensityArea(0.),
+		_dDeletedMolsPerTimestep(0.),
+		_dInvNumTimestepsSummation(0.),
+		_dMoleculeDiameter(1.0),
+		_dTransitionPlanePosY(0.0),
+		_dDensityTarget(0.0),
+		_dVolumeCV(0.0),
+		_nUpdateFreq(0),
+		_nWriteFreqRestart(0),
+		_nMaxReservoirMoleculeID(0),
+		_nNumMoleculesDeletedGlobalAlltime(0),
+		_nMovingDirection(MD_UNKNOWN),
+		_nFeedRateMethod(FRM_UNKNOWN),
+		_nZone2Method(Z2M_UNKNOWN),
+		_nNumValsSummation(0),
+		_numDeletedMolsSum(0),
+		_nDeleteNonVolatile(0)
+{
+	_dAreaXZ = global_simulation->getDomain()->getGlobalLength(0) * global_simulation->getDomain()->getGlobalLength(2);
+
+	// init restart file
+	std::stringstream fnamestream;
+	fnamestream << "MettDeamonRestart_movdir-" << (uint32_t)_nMovingDirection << ".dat";
+	std::ofstream ofs(fnamestream.str().c_str(), std::ios::out);
+	std::stringstream outputstream;
+	outputstream << "     simstep" << "   slabIndex" << "                  deltaY" << std::endl;
+	ofs << outputstream.str();
+	ofs.close();
+
+	// summation of deleted molecules
+	_listDeletedMolecules.clear();
+	_listDeletedMolecules.push_back(0);
+
+	// init identity change vector
+	uint8_t nNumComponents = global_simulation->getEnsemble()->getComponents()->size();
+	_vecChangeCompIDsFreeze.resize(nNumComponents);
+	_vecChangeCompIDsUnfreeze.resize(nNumComponents);
+	std::iota (std::begin(_vecChangeCompIDsFreeze), std::end(_vecChangeCompIDsFreeze), 0);
+	std::iota (std::begin(_vecChangeCompIDsUnfreeze), std::end(_vecChangeCompIDsUnfreeze), 0);
+
+	// density values
+	_vecDensityValues.clear();
+
+	// particle reservoir
+	_reservoir = new Reservoir(this);
```

Comment:
not released -> change to std::unique_ptr!

Reply:
Thanks for the suggestion, I changed the code as you proposed.